### PR TITLE
feat: Typography add suffix

### PR DESCRIPTION
**Pull request by zouxiaomingya**
### 🤔 This is a ...

- [x] Bug fix
- [x] New feature
- [ ] TypeScript definition update
- [ ] Component style update
- [ ] Refactoring
- [ ] Site / document update
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

1. Title bug [Example](https://codesandbox.io/s/clever-microservice-ots1f)
2. Add middle ellipsis, Typography component adds suffix attribute

### 📝 Changelog

| Language   | Changelog                                                 |
| ---------- | --------------------------------------------------------- |
| 🇺🇸 English | Typography component adds suffix attribute, Fix title bug |
| 🇨🇳 Chinese | Typography 组件增加 suffix 属性, 修复 title 的 bug        |

### ☑️ Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----
[View rendered components/typography/demo/suffix.md](https://github.com/zouxiaomingya/ant-design/blob/4.0-prepare/components/typography/demo/suffix.md)
[View rendered components/typography/index.en-US.md](https://github.com/zouxiaomingya/ant-design/blob/4.0-prepare/components/typography/index.en-US.md)
[View rendered components/typography/index.zh-CN.md](https://github.com/zouxiaomingya/ant-design/blob/4.0-prepare/components/typography/index.zh-CN.md)

### CODE REVIEW

File: components/typography/demo/suffix.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    const article =
+      'To be, or not to be, that is a question: Whether it is nobler in the mind to sufferThe slings and arrows of outrageous fortune';
+    return (
+      <div>
+        <Slider value={rows} min={1} max={10} onChange={this.onChange} />
+        <div>
+          <Paragraph ellipsis={{ rows, suffix: songForDrinking }} title={poetry + songForDrinking}>
+            <Text code strong>
+              将进酒
+            </Text>
```
Comment (zombieJ) [inline components/typography/demo/suffix.md]: 留一个英文的例子就行了

File: components/typography/demo/suffix.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            {poetry}
+          </Paragraph>
+          <Paragraph ellipsis={{ rows, expandable: true, suffix: hamlet }} title={article + hamlet}>
```
Comment (zombieJ) [inline components/typography/demo/suffix.md]: `suffix: hamlet` 直接写到这里，不用经过 `state`。让例子尽可能简单，用户易于理解。

**Comment by zombieJ** [comment]
为何 snapshot 会变？

File: components/date-picker/__tests__/__snapshots__/other.test.js.snap [hunk 3]
```diff
@@ -986,7 +986,7 @@ exports[`Picker format by locale week 1`] = `
       placeholder="请选择周"
       readonly=""
       size="12"
-      value="2000-52周"
+      value="1999-52周"
```
Comment (zombieJ) [inline components/date-picker/__tests__/__snapshots__/other.test.js.snap]: Merge `4.0-prepare`. This snapshot has updated.

File: components/typography/util.tsx [hunk 4]
```diff
@@ -90,23 +95,44 @@ export default (
...
   }
 
-  // Skip ellipsis if already match
+  render(
+    <div style={wrapperStyle}>
+      <span style={wrapperStyle}>
+        {(ellipsisStr + suffix)
+          .split('')
+          .reverse()
+          .join('')}
```
Comment (zombieJ) [inline components/typography/util.tsx]: 这段逻辑原因是什么？

Comment (zouxiaomingya) [reply components/typography/util.tsx]: 如果后缀都溢出
那么后缀也需要开始省略，但是后缀省略是需要往前面省略的。
这样倒过来还是可以复用 measureText 方法

File: components/typography/Base.tsx [hunk 5]
```diff
@@ -426,14 +429,22 @@ class Base extends React.Component<InternalBlockProps & ConfigConsumerProps, Bas
 
     // Only use js ellipsis when css ellipsis not support
     if (rows && isEllipsis && !expanded && !cssEllipsis) {
-      ariaLabel = String(children);
+      ariaLabel = title || null;
       // We move full content to outer element to avoid repeat read the content by accessibility
       textNode = (
-        <span title={String(children)} aria-hidden="true">
+        <span title={title} aria-hidden="true">
```
Comment (zombieJ) [inline components/typography/Base.tsx]: 这里有 `title` 用 `title`，如果没有但是 `children` 是 `string | number` 则用 `children`。

File: components/typography/Base.tsx [hunk 6]
```diff
@@ -285,12 +287,12 @@ class Base extends React.Component<InternalBlockProps & ConfigConsumerProps, Bas
...
   syncEllipsis() {
     const { ellipsisText, isEllipsis, expanded } = this.state;
-    const { rows } = this.getEllipsis();
+    const { rows, suffix } = this.getEllipsis();
     const { children } = this.props;
     if (!rows || rows < 0 || !this.content || expanded) return;
 
     // Do not measure if css already support ellipsis
-    if (this.canUseCSSEllipsis()) return;
+    if (this.canUseCSSEllipsis() && !suffix) return;
```
Comment (zombieJ) [inline components/typography/Base.tsx]: `suffix` 移入 `canUseCSSEllipsis` 方法内。

File: components/typography/Base.tsx [hunk 7]
```diff
@@ -426,14 +429,22 @@ class Base extends React.Component<InternalBlockProps & ConfigConsumerProps, Bas
 
     // Only use js ellipsis when css ellipsis not support
     if (rows && isEllipsis && !expanded && !cssEllipsis) {
-      ariaLabel = String(children);
+      ariaLabel = title || String(children);
       // We move full content to outer element to avoid repeat read the content by accessibility
       textNode = (
-        <span title={String(children)} aria-hidden="true">
+        <span title={title || String(children)} aria-hidden="true">
```
Comment (zombieJ) [inline components/typography/Base.tsx]: 这里和上面的 `ariaLabel` 已经一模一样了，直接放进来不用重复执行

**Comment by linxianxi** [comment]
希望后缀是 ReactNode  而不只是string

**Pull request merged**